index.bs

```diff
@@ -119,18 +120,18 @@ Overview</h2>
 
 <em>This section is not normative.</em>
 
-When spatial navigation is <a for=spatnav>active</a>,
-pressing an arrow key will either
+Using a UA-defined mechanism
+(typically arrow keys, possibly in combination with modifier keys like <code class=key>Shift</code> or <code class=key>Control</code>),
+the user may ask the User Agent to navigate in a particular direction.
+This will either
 move the focus from its current location to a new focusable item in the direction requested,
 or scroll if there is no appropriate item.
 
 More specifically,
 the User Agent will first search for visible and focusable items
 in the direction indicated
 within the current <a>spatial navigation focus container</a>
-(by default the root element, scrollable elements, and iframes,
-but other elements can be made into <a>spatial navigation focus containers</a>
-using the 'spatnav-container' property).
+(the root element, scrollable elements, or iframes)
```

Comment:
I agree we should have that property eventually, and the behavior (assuming we don't change the processsing model) is straightforward. So I am not against having it in level 1.

At the same time, we can easily add it back at any time given that it's compatible with the overall model. That makes it non-essential, and since a focused spec is good for making sure people get the point, and to limit implementation scope, I thought dropping it for now made sense.

If you'd like to keep it, I agree with putting it somewhere further down the spec, maybe after the navigation part, or even after the processing model. And in that case, yes, this sentence would need to be reverted. Should I do so, or should we keep it for level 2?

Reply:
For now, I'd like to keep it in level 1.

I think the essential part of the level 1 is the Processing Model as well as your opinion.
Processing Model uses "container" in general, and it is the vital concept of it.
There are ways to allow author-defined focus navigation heuristic and handling keys via JS API and NavigationEvent.
It also seems nice to have the author-defined container by spatnav-container.

Therefore, can we revert this sentence?

Reply:
It is possible for the author to define containers by themselves without this property, using javascript: catch the events, decide on what they want to use as the container, use `focusableAreas()` on that element, and ` spatNavSearch()` with the resulting list of candidates. So no capability is lost by dropping this property.

But it is indeed much simpler to do this with a CSS property, So I'm OK with putting it back (and revert the sentence). I would like to mark it "at risk" though, like other things that "for now we would like to keep in level 1" (as opposed to things that must absolutely be in level 1 or the spec completely changes).

Reply:
I agree with Florian here. I think we should try to avoid adding too many "convenience-helpers" for things that can be done in other ways. I'd Level 1 to only contain the smallest set of primitives needed (à l'Extensible Web).

Reply:
( In general, isn't the `navigate`-[event](https://github.com/WICG/spatial-navigation/issues/32#issuecomment-372963209) the only primitive we need to polyfill all other functionally that's been suggested in this repo? If so, I'd say implementing `navigate` should be this spec's "Level 0". )

Reply:
I realized that thoughts about core feature(the set of primitives) of SpatNav Level1 are different.
I'm with you about 'spatnav-container' can be implemented with JS.

But, while I implement the polyfill of SpatNav, I'm certain about having the customized container is convenient. Without it, all trial of spatial navigation needs to bring all candidates of the document.
Also if this is implemented with JS, it will become a burden for the performance.

Therefore I think 'spatnav-container" is essential "convenience-helper" for SpatNav Level 1.

Reply:
@hugoholgersson Could you share your thought about the primitives of Level 1?

Reply:
> Also if this is implemented with JS, it will become a burden for the performance.

In JS, only one DOM-height climb needs to be done. And that's very cheap. Blink does it all the time already, for example on every mouse movement.

```
 // Polyfilled spatnav-container: 
function handleNavigation(event) {
  if (event.target == mySpatnavContainer && !mySpatnavContainer.contains(e.relatedTarget))
    preventDefault(); // Jail navigation.
}
```

> Without it, all the trials of spatial navigation need to bring all candidates of the document.

You're right...  Every key stroke's we will check all of the document's focusables (in worst case only). I'm not sure that performance difference is big enough to motivate a new CSS property and all the logic that comes with it.

> Could you share your thought about the primitives of Level 1?

I think a navigate-event is the only primitive authors would need to experiment and come up with better constructs and smarter convenience-helpers than we can. I'd leave everything else out :)

Reply:
I think that lots of people (myself included) see a level 0 hiding inside this level 1 as a subset of what's in there. The problem is that we all seem to see a different subset.

For now, I'm dealing with this with marking possibly non essential parts "at-risk"